### ISSUE TITLE
Question regarding KL Divergence gradient flow in `AdaptiveLayerLoss` (edit: and possible issue with layers weighting schema)

### ISSUE THREAD

**Issue by bobox2997** [UKPLab/sentence-transformers#3757]
Hi @tomaarsen ,

I’ve been looking through the implementation of `AdaptiveLayerLoss` and had a question regarding the KL divergence computation between the intermediate layers and the final layer. I wanted to check if the current behavior is intended or if I might be misunderstanding a part of the implementation.

### Observation: 
In the `forward` method, the loss calculates `final_embeddings` from the full model to serve as a target for the intermediate layers:

```if self.kl_temperature > 0:    final_embeddings = forward_decorator.get_embeddings()    final_embeddings = F.softmax(final_embeddings / self.kl_temperature, dim=-1)```

Later, this is used as the `target` in `F.kl_div`:

```kl_div_loss = F.kl_div(   F.log_softmax(embeddings / self.kl_temperature, dim=-1),   final_embeddings,   reduction="batchmean",)```

(this seems to be common to both  `AdaptiveLayerLoss` and `Matryoshka2dLoss`)

### My Question: 
currently, `final_embeddings` is not detached from the computation graph. from my understanding, in PyTorch, this means that during the backward pass, gradients will flow not only into the intermediate `embeddings` (the "student", if we frame that as a distillation approach) but also back into the `final_embeddings` (the "teacher").

this creates a bidirectional optimization where the final layer is penalized for being different from the intermediate layers. In most distillation setups, ad far as i know, the teacher's output is typically detached, for obvious reasons. (still, one could argue if this really apply to this context of "self-distillation".

**Is it the intentional design of `AdaptiveLayerLoss` to allow the intermediate layers to influence the final layer's representations/optimization via this KL loss?**

or should the final layer be treated as a fixed reference point, like so?
```final_embeddings = F.softmax(final_embeddings / self.kl_temperature, dim=-1).detach()```

I'm curious to hear the thoughts of the maintainers on whether this bidirectional gradient flow is a feature of the 2DMSE approach or if it's something that might need a `.detach()`.

...a side (speculative) note:

interestingly, the recent 'Starbucks' paper (Zhuang et al., 2024, https://arxiv.org/pdf/2410.13230v1) notes that 2DMSE/ESE training can 'significantly hurt final-layer performance.' Given that they used the sentence-transformers implementation (*'For consistency, we implemented using the sentence transformers library which already included 2DMSE baseline implementations', section 4.2.1*), it's possible that the 'interference' they observed is actually the result of (or has some contribution from) this bidirectional gradient flow caused by the missing .detach() ?

btw, thanks for the great library!

Piero Boboli

*edit: fixed formatting/typo*
*edit: paper reference to v3: https://arxiv.org/pdf/2410.13230v3*

**Comment by bobox2997** [comment]
I'm sorry, i previusly forgot, but as I am here... out of curiosity, what is the rationale for the current layers weight scaling logic? is this intended?

the current implementation use ```loss = loss + layer_loss / (1 + layer_idx) / len(layer_indices) * self.prior_layers_weight```, but it seems a little counter-intuitive.

I couldn't find a reference to this specific discounting logic in the original 2DMSE (v1) or Espresso (v2) papers. 

Furthermore, the Starbucks paper (Zhuang et al., 2024) defines its objective in Algorithm 1 as a uniform summation across configurations. If the authors used the ST 2DMSE implementation 'as-is' for their baseline but used uniform weighting for their proposed method, that alone could explain a portion of the performance gap they observed.

if "Starbucks: Improved Training for 2D Matryoshka Embeddings", Zhuang et al., 2024, (https://arxiv.org/pdf/2410.13230v1) used Sentence Transformers 2DMSE loss as baseline, and removed this scaling for their propose method, an ablation should be considered. 

I'm tagging the authors of the original research, @SeanLee97 and @ArvinZhuang, as they might be interested in whether these implementation details in the library could explain the 'gradient conflict' and final-layer performance issues discussed in their work (@guidozuc / @ielab).

**Comment by bobox2997** [comment]
Regardless of the original intent behind these design choices, it might be beneficial to make these behaviors configurable in `AdaptiveLayerLoss` (and by extension `Matryoshka2dLoss`), or at least explicitly document the current gradient flow and layer scaling mechanisms.


@tomaarsen I just found out that you mentioned this likely being a bracket typo in an old issue I opened for another reason, but I completely forgot about that:
https://github.com/huggingface/sentence-transformers/issues/3329#issuecomment-2871974825

I could open a PR about that, but I have some questions:

For the layer scaling, should I implement the originally intended deep-weighted scaling `((1 + layer_idx) / len(layer_indices))`, or should I use a uniform summation across all layers, which as I understand is what the original 2DMSE/ESE paper and the recent Starbucks paper (Starbucks, Zhuang et al 2024-2025) use? 

should I include the detach fix?

**Comment by bobox2997** [comment]
@tomaarsen, gently bumping this as a week passed since I opened the issue. (@ielab , for context)
if you are busy I can directly make the PR, just react to this message.

**Comment by tomaarsen** [comment]
Hello!

Please do, I haven't yet had time to look into this in more detail. I appreciate it!

- Tom Aarsen

**Comment by wshuai190** [comment]
Hi @bobox2997, thanks for the careful read. Quick answers:

1. **Detach**: agreed, missing `.detach()` doesn't look like a norm (sorry did not notice that before), standard self-distillation uses a stop-gradient teacher, and without it KL gradients also pull the final layer toward the intermediate ones.

2. **Layer weighting**: the original [v2/ESE](https://arxiv.org/pdf/2402.14776v2) Eq. 2 uses **log** weighting `w_i = 1/(1+ln(i))` with the last layer unweighted. The library's `1/(1+layer_idx)/N` is **linear** decay which I'm guessing is a mimic of the original V2?

3. **What Starbucks used**: our faithful 2DMSE reproduction (with log weighting + detached KL) lives in [ielab/2DMSE-Reproduce](https://github.com/ielab/2DMSE-Reproduce) (paper: [arxiv 2411.17299](https://arxiv.org/pdf/2411.17299)). In the Starbucks paper itself we just used `Matryoshka2dLoss` from sentence-transformers as the 2DMSE baseline for simplicity, so that STS baseline does inherit the library's linear weighting and non-detached KL, and your concern about the comparison being confounded is fair. Our [Starbucks retrieval code](https://github.com/ielab/Starbucks/blob/main/retrieval/modeling.py#L125) is fine on both counts (explicit `.detach()` + uniform per-layer weight).

**Comment by bobox2997** [comment]
Hi @wshuai190 , thank you for the detailed answer, apreciate that!

>the original [v2/ESE](https://arxiv.org/pdf/2402.14776v2) Eq. 2 uses log weighting w_i = 1/(1+ln(i)) with the last layer unweighted.

oh i'm honestly sorry, i somehow missed that entirely while reading the paper, didn't mean to spread a false information. 

i'll give [arxiv 2411.17299](https://arxiv.org/pdf/2411.17299)) a proper read, thanks! 

out of curiosity, do you believe the missing detach could have had a relevant effect on the training dynamics?

@tomaarsen in the PR, other than adding the detach, i'm going to add a flag to select between the log scaling and uniform weighting (in addition to the current implementation)...  which one should i keep as default?

**Comment by tomaarsen** [comment]
@bobox2997 Apologies for the delay. I'm okay with updating the default weighting to whatever is recommended in the most recent works, even if that means that older training scripts will have slightly updated results.

- Tom Aarsen

**Comment by Kaif10** [comment]
Is anyone actively working on this? It's been a few weeks since the discussion, and I don't see a linked PR yet — @bobox2997, if you're still planning to take it, I don't want to step on your work.

If it's open, I'm happy to send a small, tightly-scoped PR covering the two points that seem settled here:
1. `.detach()` the teacher (`final_embeddings`) before it's used as the `F.kl_div` target, so gradients don't flow back into the final layer (the standard self-distillation setup).
2. The bracket/precedence fix in the per-layer weighting that @tomaarsen flagged as a likely typo.

I'd leave the broader question of the default weighting scheme (uniform vs. distance-weighted) out of this PR, since that's a design choice you were still discussing — happy to follow whatever you land on there. Tests would assert the teacher receives no gradient and that the weighting matches the intended formula.

**Comment by bobox2997** [comment]
@Kaif10 hey, I'm sorry if i didn't follow up on this. I made added a fix but didn't submit it, unfortunately life happens, I had some issue with my uni and I hadn't time to commit.
from the 30 of July, however, I'll have some free time and I'll submit it.

if you are in a hurry and need the repo fixed asap, you can proceed, ofc.


again, I'm sorry

**Comment by Kaif10** [comment]
Thanks @bobox2997, and no worries at all — hope the uni stuff eases up. Since you mentioned it was fine to proceed if it helps, I've opened #3880 with the two scoped fixes we all converged on (detach the KL teacher + the bracket-typo `((1 + layer_idx) / len(layer_indices))` that @tomaarsen identified), plus regression tests. I've credited your report and the discussion in the PR. I deliberately left the weighting-*scheme* question (uniform vs. log) out of it since that's still open — happy to defer to whatever you and the maintainers decide there.

**Comment by bobox2997** [comment]
@Kaif10, well...

>Since you mentioned it was fine to proceed if it helps

to be honest, I mentioned it was fine to proceed (from my point ofc, I'm not the maintainer, but since you asked to me...) if you couldn't wait 5 days and needed the code fixed on the main immediately, as I already had updated the code and tested, I just needed to submit it.

what does "if it helps" mean? 

I mean, i have no rights on anything here, I failed to follow up on that, you could have simply submitted the PR, but what's the point on asking then submitting "if it helps"? 

but whatever help the repo and the community, I guess... I'm not here to argue about a two line PR on a public repo.

**Comment by Kaif10** [comment]
Fair point, and apologies — you're right that I could have just opened it; asking and then proceeding was clumsy on my part. The diagnosis here was yours and @wshuai190's, and I've credited that in the PR. Happy to defer to you and @tomaarsen on the weighting-scheme question. Thanks for being gracious about it.

**Comment by tomaarsen** [comment]
Reopening for the second half: the configurable layer weighting (#3880 covered the KL teacher detach).
